bin/get-video-s3-acls.py

```diff
@@ -0,0 +1,377 @@
+        "dg.published AS gloss_public, "
+        "vg.is_public AS video_public, "
+        "vg.id AS video_id, "
+        "vg.videofile AS video_key "
+        "FROM dictionary_gloss AS dg JOIN video_glossvideo AS vg ON vg.gloss_id = dg.id"
+        ") TO STDOUT WITH (FORMAT CSV, DELIMITER '|')",
+    )
+
+    # Separate the NZSL db columns
+    # Write them to a dictionary, so we can do fast operations
+    for rawl in result.stdout.split("\n"):
+        rawl = rawl.strip()
+        if not rawl:
+            continue
+        [
+            gloss_id,
+            gloss_idgloss,
+            gloss_created_at,
+            gloss_public,
+            video_public,
+            video_id,
+            video_key,
+        ] = rawl.split("|")
+
+        this_nzsl_raw_keys_dict[video_key] = [
+            gloss_idgloss.replace(CSV_DELIMITER, ""),
+            gloss_created_at,
+            gloss_id,
+            video_id,
+            gloss_public.lower() == "t",
+            video_public.lower() == "t",
+        ]
+
+    print(
+        f"{len(this_nzsl_raw_keys_dict)} rows retrieved",
+        file=sys.stderr,
+    )
+
+    return this_nzsl_raw_keys_dict
+
+
+# Get all keys from AWS S3
+def get_s3_bucket_raw_keys_list(s3_bucket=AWS_S3_BUCKET):
+    print(f"Getting raw AWS S3 keys recursively ({s3_bucket}) ...", file=sys.stderr)
+    result = aws_cli(
+        [
+            "s3",
+            "ls",
+            f"s3://{s3_bucket}",
+            "--recursive",
+        ],
+    )
+
+    # Separate out just the key from date, time, size, key
+    this_s3_bucket_raw_keys_list = []
+    for line in result.stdout.split("\n"):
+        if line:
+            this_s3_bucket_raw_keys_list.append(re.split(r"\s+", line, 3)[3])
+
+    print(
+        f"{len(this_s3_bucket_raw_keys_list)} rows retrieved",
+        file=sys.stderr,
+    )
+
+    return this_s3_bucket_raw_keys_list
+
+
+# Get the keys present and absent across NZSL Signbank and S3, to dictionary
+def create_all_keys_dict(this_nzsl_raw_keys_dict, this_s3_bucket_raw_keys_list):
+    print(
+        "Getting keys present and absent across NZSL Signbank and S3 ...",
+        file=sys.stderr,
+    )
+    this_all_keys_dict = {}
+
+    # Find S3 keys that are present in NZSL, or absent
+    for video_key in this_s3_bucket_raw_keys_list:
+        dict_row = this_nzsl_raw_keys_dict.get(video_key, None)
+        if dict_row:
+            # Split out for readability
+            [
+                gloss_idgloss,
+                gloss_created_at,
+                gloss_id,
+                video_id,
+                gloss_public,
+                video_public,
+            ] = dict_row
+
+            this_all_keys_dict[video_key] = [
+                True,  # NZSL PRESENT
+                True,  # S3 PRESENT
+                gloss_idgloss,
+                gloss_created_at,
+                gloss_id,
+                video_id,
+                gloss_public,
+                video_public,
+            ]
```

Comment:
suggestion: similar to my comments in previous reviews, we should be able to just add the contents of the row into this new array

```suggestion
            this_all_keys_dict[video_key] = [
                True,  # NZSL PRESENT
                True,  # S3 PRESENT
            ] + dict_row
```

That way this function does not need to know about the specific cells in the row

Reply:
Yes but the human coming along to debug it later, does! 😄

Reply:
The order shouldn't be needed here if we're just passing through information, which is exactly what we're doing and why this is worth doing - the current layout means you've got 3+ places where the order could be changed meaning there's more surface area for bugs

Reply:
It does mean matching the column headers up is trickier too. But I'll compress this and see what it looks like.

Reply:
Done!